File: EPIJudge/epi_judge_python/binomial_coefficients.py
from test_framework import generic_test
import math
from collections import defaultdict

def fact(coef, n, op):
    for k in range(2, int(math.sqrt(n) + 1)):
        while n and n % k == 0:
            coef[k] += op
            n /= k
    if n:
        coef[n] += op


def compute_binomial_coefficient(n: int, k: int) -> int:
    coef = defaultdict(int)

    den = max(n - k, k)
    num = n - den

    for i in range(num + 1, n + 1):
        fact(coef, i, 1)

    for i in range(2, den + 1):
        fact(coef, i, -1)

    mult = 1
    for b, e in coef.items():
        mult *= int(pow(b, e))
    return mult


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('binomial_coefficients.py',
                                       'binomial_coefficients.tsv',
                                       compute_binomial_coefficient))

File: EPIJudge/epi_judge_cpp/binomial_coefficients.cc
#include "test_framework/generic_test.h"

void fact(unordered_map<int, int> &coeff, int t, int op) {
    for (int j = 2; j <= sqrt(t); j += 1) {
        while (t % j == 0 && t) {
            t /= j;
            coeff[j] += op;
        }
    }
    if (t) {
        coeff[t] += op;
    }
}

int ComputeBinomialCoefficient(int n, int k) {
    unordered_map<int, int> coeff;
    int den = min(n - k, k);
    int num = n - den + 1;
    for (int i = num; i <= n; i += 1) {
        fact(coeff, i, 1);
    }
    for (int i = 2; i <= den; i += 1) {
        fact(coeff, i, -1);
    }
    int mult = 1;
    for (auto & it : coeff) {
        mult *= pow(it.first, it.second);
    }
    return mult;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"n", "k"};
  return GenericTestMain(
      args, "binomial_coefficients.cc", "binomial_coefficients.tsv",
      &ComputeBinomialCoefficient, DefaultComparator{}, param_names);
}
